### PR TITLE
Issue/1482 step3 handle discovery error changes

### LINKED ISSUE #1482: Jetpack sign-in with self-hosted credentials - Master checklist
This is the master epic for the changes required to accomplish jetpack sign-in with self-hosted credentials.

#### Changes in FluxC:
- [#1402](https://github.com/wordpress-mobile/WordPress-FluxC-Android/issues/1402) - Fetch `jetpack_user_email` param from from the `xmlrpc wp.getOptions` endpoint and cache locally. (PR: [#1403](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1403)) 

#### Changes in Login library:
- [x] Step 1: Fetch Jetpack e-mail address for flowing through Magic Links: (PR: #1483 )
  - [x] Revert existing logic of validating site url from `FETCH_CONNECT_SITE_INFO` endpoint and use the `FETCH_WPCOM_SITE_BY_URL` endpoint and initiate the discovery process.
  - [x] Add new button to email screen to `Login via site credentials`.
  - [x] When site credentials are clicked:
     - [x] If jetpack is installed/active/connected: call the `wp.getOptions` endpoint to fetch the `jetpack_user_email` and store in `SiteModel` and rRedirect to the request magic link screen.
     - [x] If jetpack NOT installed/active/connected: Redirect to Jetpack required screen.
- [x] Step 2: Initiate verification email for the jetpack connected email address. (PR: #1498 )
- [ ] Step 3: Add Track events for the new buttons/views.
- [x] Step 4: Handle discovery error scenarios for Woo. (PR: #1515 )
     - [x] This includes creating a base class that handles the discovery process and provides an option for the calling class to handle the errors as they see fit.
     - [x] One of the tasks include porting over changes from the `LoginSiteAddressFragment` to utilise this base class.

Screenshot on 2019-10-21 at 14_27_17

### PR DESCRIPTION
**Pull request by anitaa1990**
Partially fixes #1482. This PR completes the third step of the Jetpack sign in with self credentials as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1482).

### Changes:
- ~~If a discovery error takes place for a site,~~
- Further testing revealed that some errors such as XMLRPC_FORBIDDEN, XMLRPC_BLOCKED and MISSING_XMLRPC_METHOD errors can occur even with Jetpack available on the site. Furthermore, the `CONNECT_SITE_INFO` API returns that Jetpack does not exist, even when Jetpack is available and connected on the site. In order to address this issue, the logic for discovery errors has being modified as follows:
  - For HTTP Auth error: HTTP Auth is not supported by Jetpack so we can safely assume that this error would only occur if Jetpack is not connected on a site. So we redirect the user to the Jetpack required screen. There is a possibility that user installs Jetpack while on this screen and clicks on `Try again`. In order to avoid confusion and displaying No jetpack available when it is, we are using the `CONNECT_SITE_INFO` API to verify if Jetpack is installed/active/connected. 
    - If Jetpack is connected, redirect them back to the site credentials screen.
    - If Jetpack is still not connected, display an error snackbar asking them to install Jetpack.
  - All other errors that can occur: Regardless of whether Jetpack is available or note, redirect them to the discovery error screen. 
    - Added a new fragment called `LoginDiscoveryErrorFragment` that is displayed when a discovery error takes place for a site that has jetpack available.
     - Clicking on `Sign in with WordPress.com` redirects  to the email screen.
     - Clicking on `Read our troubleshooting tips` redirects to this [link](https://jetpack.com/support/getting-started-with-jetpack/troubleshooting-tips/) inside a webview.
     - Clicking on `Try again` redirects to the site credentials screen.
- Ported over discovery initiation from the `LoginSiteAddressFragment` to the `LoginBaseDiscoveryFragment`.

### Notes:
- The are TODO's added to the PR to add events to the button clicks. This will take place in another PR.
- Note that this PR is to be merged to a feature branch.

#### Logging into a site with HTTP AUTH required when Jetpack is NOT available:
(LEFT: When jetpack is still not installed and user clicks on Try AGAIN button)
(RIGHT: When jetpack is installed when user is on the screen and clicks on TRY AGAIN button)
 . 

#### Different discovery error messages:
(LEFT: HTTP AUTH required . RIGHT: SSL certificate needed)
 . 

(LEFT: XMLRPC errors . RIGHT: All other generic errors)
 . 

##### Scenario I: Login to a site where the XMLRPC file is renamed.  
(This scenario is applicable to the following discovery errors: `MISSING_XMLRPC_METHOD`, `XMLRPC_BLOCKED`, `XMLRPC_FORBIDDEN`, `GENERIC_ERROR`, `NO_SITE_FOUND`)
- Login to anitaamurthy.com in `wp-admin`.
- Install, activate and connect Jetpack.
- Open the app and enter the above site address and click on `Next`.
- Verify that you are redirected to the Sign in with Jetpack screen (even though Jetpack is installed on the site). This is because the XMLRPC file is renamed for this site and cannot be found so even if Jetpack is available, the API response returns false for the fields `	hasJetpack: true`, `isJetpackActive: false`, `isJetpackConnected: false`.
- Click on Sign in in the app.
- Click on `Login with site credentials` and enter username + password.
- Verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access the XMLRPC file on your site. You will need to reach out to your host to resolve this.`.
- Verify that clicking on `Try again` redirects to the Username + password screen.
- Verify that clicking on `Read our troubleshooting tips` redirects to a webview from this link.
- Verify that clicking on `Help` opens the help page.
- Verify that clicking in `Sign in with Wordpress.com` redirects to the Email screen and you are able to login successfully to the app using wp.com email.

##### Scenario II: Login to a site with SSL Certificate required error:
  (This scenario is the same as the above scenario but with different discovery error type: `ERRONEOUS_SSL_CERTIFICATE`)
- Login to anitaastestwpsite2.com in `wp-admin`.
- Install, activate and connect Jetpack.
- Open the app and enter the above site address and click on `Next`.
- Click on `Login with site credentials` and enter username + password.
- Verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access your site because of a problem with the SSL Certificate. You will need to reach out to your host to resolve this.`.
- Verify that clicking on `Try again` redirects to the Username + password screen.
- Verify that clicking on `Read our troubleshooting tips` redirects to a webview from this link.
- Verify that clicking on `Help` opens the help page.
- Verify that clicking in `Sign in with Wordpress.com` redirects to the Email screen and you are able to login successfully to the app using wp.com email.

##### Scenario III: Login to a site where HTTP AUTH plugin is installed and activated.  
- Login to anitaastestwpsite.com in `wp-admin`.
- Uninstall Jetpack from the Plugins section.
- Activate `HTTP Auth` plugin from the same page. This enables HTTP Auth to the site.
- Open the app and enter the above site address and click on `Next`.
- Verify that you are redirected to the Sign in with Jetpack screen (since Jetpack is not active on the site).
- Click on Sign in in the app.
- Click on `Login with site credentials` and enter username + password.
- Verify you are redirected to the Jetpack required screen. 
- Click on Try again.
- Verify that an error snackbar is displayed stating that Jetpack is not active on the site.
- Go back to the Plugins page in `wp-admin` and deactivate `HTTP Auth` before activating and connecting Jetpack (This is because Jetpack does not support HTTP Auth plugin and you will only be able to connect Jetpack if HTTP Auth plugin is deactivated).
- Go back to the app and click on Try again.
- Verify that you are redirected back to the site credentials screen.
- Click on `Next` and verify that you are successfully redirected to the Magic link screen.

##### Scenario IV: Login to a site where the XMLRPC request is blocked.
- Login to any test site in wp-admin.
- Install the plugin `Disable XML-RPC`.
- Open the app and enter the above site address and click on Next.
- If Jetpack is not active on the site, verify that: 
     - You are redirected to the Sign in with Jetpack screen.
     - Click on Sign in in the app.
     - Click on Login with site credentials and enter username + password.
- If Jetpack is active on the site, verify that: 
     - You are redirected to the username password screen.
     - Click on Login with site credentials and enter username + password.
- In either case, verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access the XMLRPC file on your site. You will need to reach out to your host to resolve this.`.  

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

@AmandaRiu, I assigned the PR to you since you have reviewed the last two 😄 Could you trouble you for one more review please? Thank you!!

### CODE REVIEW

**Review by AmandaRiu — COMMENTED** [review]
@anitaa1990 finished the first review round and left a code comment. I tried testing through the scenarios using our test sites page but couldn't get to the `FragmentLoginDiscoveryError` view. I tried to sign in with a WP test site that had HTTP AUTH required, but couldn't get past the first screen because `CONNECTED-SITE-INFO` was flagging it as a site that was not a WordPress site. Logging in with a site that required an SSL cert didn't give me this screen either - it would just drop me in the "Jetpack Required" screen. 



Do you have some tests sites I can use?

File: WooCommerce/src/main/res/layout/fragment_login_discovery_error.xml [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+                        style="@style/Woo.Login.DiscoveryError.Message"
+                        android:layout_width="0dp"
+                        android:layout_height="wrap_content"
+                        android:layout_margin="@dimen/default_padding"
+                        android:paddingStart="@dimen/card_padding_start"
+                        android:paddingEnd="@dimen/card_padding_end"
+                        android:text="@string/login_troubleshooting_tips"
+                        app:layout_constraintBottom_toBottomOf="parent"
+                        app:layout_constraintLeft_toRightOf="@+id/troubleshoot_option_icon"
+                        app:layout_constraintTop_toTopOf="parent" />
```
[the remark this answers is not in the document]
Comment (anitaa1990) [reply WooCommerce/src/main/res/layout/fragment_login_discovery_error.xml]: Thanks for the tip 😄 Done in d75b227

**Comment by anitaa1990** [comment]
Thanks for the quick review @AmandaRiu. 

I've invited you to two of my test sites. The invites are sent to your official email account. Do let me know if you face any issues. I've also updated the PR description with the testing steps and explanation.

PR ready for another round 👍

**Comment by AmandaRiu** [comment]
@anitaa1990 Thank you for the test accounts and testing steps! All of them worked great except for this one: **Scenario II: Login to a site with missing rsd tags**. With that scenario I never get the "Jetpack required" view. It allows me to get all the way to the "Not a woocommerce site" view.

**Comment by anitaa1990** [comment]
Hey @amandaRiu, thanks for testing! 

> All of them worked great except for this one: Scenario II. With that scenario I never get the "Jetpack required" view. It allows me to get all the way to the "Not a woocommerce site" view.

Ah, yes! I was testing SSL certification with that test site and had to revoke the changes I made that caused this error. I forgot yo update you 🤦‍♀️  I’ve set an invalid certificate to the site now so you should be able to see the below behaviour when you try it now. The steps to reproduce this are updated in the PR as well under `Scenario II: Login to a site with SSL Certificate required error`.


I also faced another error when trying to login with a site that has the `Disable XMLRPC` plugin. This plugin blocks the XMLRPC calls but interestingly, the discovery API returns a successful response. But when trying to access the site details using `wp.getUserBlogs` API endpoint, results in an error. I’ve modified this behaviour in the app as well in `d58f81f ` and updated the test scenario to include this test scenario as well. To reproduce this error:
- Login to any test site in wp-admin.
- Install the plugin `Disable XML-RPC`.
- Open the app and enter the above site address and click on Next.
- If Jetpack is not active on the site, verify that: 
     - You are redirected to the Sign in with Jetpack screen.
     - Click on Sign in in the app.
     - Click on Login with site credentials and enter username + password.
- If Jetpack is active on the site, verify that: 
     - You are redirected to the username password screen.
     - Click on Login with site credentials and enter username + password.
- In either case, verify that you are redirected to the discovery error screen and the error message displayed is `We were unable to access the XMLRPC file on your site. You will need to reach out to your host to resolve this.`.

Let me know if you face any other issues! PR ready for another round 👍

**Review by AmandaRiu — APPROVED** [review]
Brilliant work @anitaa1990! I've run through the new test scenarios and they were all successful! 🎆  Approving this PR :shipit:

**Pull request merged**